File: GolangSolutions/SWM/M3134_找出唯一性数组的中位数.go
/*
	@author Hyper
	@date 2024/08/27
	@file M3134_找出唯一性数组的中位数.go
*/

package SWM

func medianOfUniquenessArray(nums []int) int {
	n := len(nums)
	tot := (n*(n+1)/2 + 1) / 2
	var check func(mid int) bool
	check = func(mid int) bool {
		mp := make(map[int]int)
		cnt := 0
		for l, r := 0, 0; r < n; r++ {
			mp[nums[r]]++
			for len(mp) > mid {
				p := nums[l]
				l++
				mp[p]--
				if mp[p] == 0 {
					delete(mp, p)
				}
			}
			cnt += r - l + 1
			if cnt >= tot {
				return true
			}
		}

		return false
	}
	low, high := 0, n
	for low+1 < high {
		mid := low + ((high - low) >> 1)
		if check(mid) {
			high = mid
		} else {
			low = mid
		}
	}

	return high
}

File: CPPSolutions/LeetCode/SWM/M3134_找出唯一性数组的中位数.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * @author Hyper
 * @date 2024/08/27
 * @file M3134_找出唯一性数组的中位数.cpp
 *
 * 注意这题tot算的时候1e10正好会炸，需要强转一下
 */
#define LL long long
#define N 100010
int freq[N];

class Solution {
public:
    int medianOfUniquenessArray(vector<int> &nums) {
        int n = nums.size();
        LL tot = ((LL) n * (n + 1) / 2 + 1) / 2;
        auto check = [&](int mid) -> bool {
            LL cnt = 0;
            memset(freq, 0, sizeof(freq));
            for (int l = 0, r = 0, len = 0; r < n; r++) {
                if (++freq[nums[r]] == 1) len++;
                while (len > mid)
                    if (--freq[nums[l++]] == 0)
                        len--;
                cnt += r - l + 1;
                if (cnt >= tot) return true;
            }

            return false;
        };
        int low = 0, high = n;
        while (low + 1 < high) {
            int mid = low + ((high - low) >> 1);
            (check(mid) ? high : low) = mid;
        }

        return high;
    }
};
